### ISSUE TITLE
adc_digi_read_bytes() messes the order of channels between read chuncks  (IDFGH-9437)

### ISSUE THREAD

**Issue by ves011** [espressif/esp-idf#10804]
### Answers checklist.

- [X] I have read the documentation [ESP-IDF Programming Guide](https://docs.espressif.com/projects/esp-idf/en/latest/) and the issue is not addressed there.
- [X] I have updated my IDF branch (master or release) to the latest version and checked that the issue is present there.
- [X] I have searched the issue tracker for a similar issue and not found a similar issue.

### IDF version.

v4.4.3

### Operating System used.

Windows

### How did you build your project?

Eclipse IDE

### If you are using Windows, please specify command line type.

None

### Development Kit.

ESP32-S3

### Power Supply used.

USB

### What is the expected behavior?

adc_digi_read_bytes() for multiple channels and given number of samples shall return the channel values in the same order  in each data chunk returned by function call.
e.g. for 2 channel ADC1 only, i expect 
chunk1 = **data_chn_a**, **data_chn_b**,  **data_chn_a**, **data_chn_b**,  **data_chn_a**, **data_chn_b**, ...
if chunk1 ends with **data_chn_a**, chunk2 shall start with **data_chn_b**,  data_chn_a, data_chn_b,  data_chn_a, data_chn_b, ...
if chunk1 ends with **data_chn_b**, chunk2 shall start with **data_chn_a**, data_chn_b,  data_chn_a, data_chn_b,  data_chn_a, ...
In this way the assembled buffer will not miss any sample.

### What is the actual behavior?

Instead the following chunk start randomly with data_chn_a or data_chn_a messing the order in the assembled buffer.
Maybe some samples are lost, which for my case is very critical

### Steps to reproduce.

this is a sample code to reporoduce
```
typedef union
    {
    uint8_t result[NR_SAMPLES * ADC_RESULT_BYTE];
    adc_digi_output_data_t presult[NR_SAMPLES];
    } res_t;

#define    SENSOR_ADC_CHANNEL        (4)
#define CURRENT_ADC_CHANNEL        (3)
static uint16_t adc1_chan_mask = BIT(SENSOR_ADC_CHANNEL) | BIT(CURRENT_ADC_CHANNEL);
static uint16_t adc2_chan_mask = BIT(0);
static adc_channel_t channel[] = {CURRENT_ADC_CHANNEL, SENSOR_ADC_CHANNEL};
static void continuous_adc_init(uint16_t adc1_chan_mask, uint16_t adc2_chan_mask, adc_channel_t *channel, uint8_t channel_num)
    {
    adc_digi_init_config_t adc_dma_config =
        {
        .max_store_buf_size = NR_SAMPLES * ADC_RESULT_BYTE,
        .conv_num_each_intr = NR_SAMPLES,
        .adc1_chan_mask = adc1_chan_mask,
        .adc2_chan_mask = adc2_chan_mask,
        };
    ESP_ERROR_CHECK(adc_digi_initialize(&adc_dma_config));

    adc_digi_configuration_t dig_cfg =
        {
        .conv_limit_en = ADC_CONV_LIMIT_EN,
        .conv_limit_num = 0,
        //.sample_freq_hz = 20 * 1000,
        .sample_freq_hz = 4 * 1000,
        .conv_mode = ADC_CONV_SINGLE_UNIT_1,
        .format = ADC_DIGI_OUTPUT_FORMAT_TYPE2,
        };

    adc_digi_pattern_config_t adc_pattern[4] = {0};
    dig_cfg.pattern_num = channel_num;
    for (int i = 0; i < channel_num; i++)
        {
        uint8_t unit = GET_UNIT(channel[i]);
        uint8_t ch = channel[i] & 0x7;
        adc_pattern[i].atten = ADC_ATTEN_DB_11;
        adc_pattern[i].channel = ch;
        adc_pattern[i].unit = unit;
        adc_pattern[i].bit_width = SOC_ADC_DIGI_MAX_BITWIDTH;

        //ESP_LOGI(TAG, "adc_pattern[%d].atten is :%x", i, adc_pattern[i].atten);
        //ESP_LOGI(TAG, "adc_pattern[%d].channel is :%x", i, adc_pattern[i].channel);
        //ESP_LOGI(TAG, "adc_pattern[%d].unit is :%x", i, adc_pattern[i].unit);
        }
    dig_cfg.adc_pattern = adc_pattern;
    ESP_ERROR_CHECK(adc_digi_controller_configure(&dig_cfg));
    }

void app_main()
     {
...

    continuous_adc_init(adc1_chan_mask, adc2_chan_mask, channel, sizeof(channel) / sizeof(adc_channel_t));
...
       res_t res;
       index = 0;
       ret = ESP_OK;
       uint32_t b2read = NR_SAMPLES * ADC_RESULT_BYTE;
       adc_digi_start();
       while(index < NR_SAMPLES * ADC_RESULT_BYTE && ret == ESP_OK)
           {
           ret = adc_digi_read_bytes(res.result + index, b2read, &ret_num, 1000);
           if (ret == ESP_ERR_TIMEOUT || ret == ESP_ERR_INVALID_STATE)
               break;
           ESP_LOGI(TAG, "read chunk %d %d", index, ret_num);
           index += ret_num;
           b2read -= ret_num;
           }
       adc_digi_stop();
...
}
```
If you look in res.presult at chunk boundary from time to time you'll see the order of values is changed.

**Comment by igrr** [comment]
This might be the same issue as https://github.com/espressif/esp-idf/issues/10636.

**Comment by ves011** [comment]
hi @Icarus113,
Thx for taking your time to look into this issue!
Here is a summary of my results after 24 cycles with NR_SAMPLES = 400

- each cycle has 5 chunks with different length
(summary extracted from the log)
Line   369: I (17:01:57.648) ADC OP: read chunk 0 208
	Line   370: I (17:01:57.672) ADC OP: read chunk 208 388
	Line   371: I (17:01:57.697) ADC OP: read chunk 596 400
	Line   372: I (17:01:57.722) ADC OP: read chunk 996 400
	Line   373: I (17:01:57.747) ADC OP: read chunk 1396 204
	Line   775: I (17:02:00.018) ADC OP: read chunk 0 196
	Line   776: I (17:02:00.042) ADC OP: read chunk 196 388
	Line   777: I (17:02:00.067) ADC OP: read chunk 584 400
	Line   778: I (17:02:00.092) ADC OP: read chunk 984 400
	Line   779: I (17:02:00.117) ADC OP: read chunk 1384 216
	Line  1181: I (17:02:02.378) ADC OP: read chunk 0 184
	Line  1182: I (17:02:02.402) ADC OP: read chunk 184 384
	Line  1183: I (17:02:02.427) ADC OP: read chunk 568 400
	Line  1184: I (17:02:02.452) ADC OP: read chunk 968 400
	Line  1185: I (17:02:02.477) ADC OP: read chunk 1368 232
	Line  1587: I (17:02:04.738) ADC OP: read chunk 0 168
	Line  1588: I (17:02:04.762) ADC OP: read chunk 168 388
	Line  1589: I (17:02:04.787) ADC OP: read chunk 556 400
	Line  1590: I (17:02:04.812) ADC OP: read chunk 956 400
	Line  1591: I (17:02:04.837) ADC OP: read chunk 1356 244
	Line  1993: I (17:02:07.088) ADC OP: read chunk 0 156
	Line  1994: I (17:02:07.112) ADC OP: read chunk 156 388
	Line  1995: I (17:02:07.137) ADC OP: read chunk 544 400
	Line  1996: I (17:02:07.162) ADC OP: read chunk 944 400
	Line  1997: I (17:02:07.187) ADC OP: read chunk 1344 256
	Line  2403: I (17:02:09.448) ADC OP: read chunk 0 144
	Line  2404: I (17:02:09.472) ADC OP: read chunk 144 388
	Line  2405: I (17:02:09.497) ADC OP: read chunk 532 400
	Line  2406: I (17:02:09.522) ADC OP: read chunk 932 400
	Line  2407: I (17:02:09.547) ADC OP: read chunk 1332 268
	Line  2809: I (17:02:11.808) ADC OP: read chunk 0 132
	Line  2810: I (17:02:11.832) ADC OP: read chunk 132 384
	Line  2811: I (17:02:11.857) ADC OP: read chunk 516 400
	Line  2812: I (17:02:11.882) ADC OP: read chunk 916 400
	Line  2813: I (17:02:11.907) ADC OP: read chunk 1316 284
	Line  3215: I (17:02:14.168) ADC OP: read chunk 0 116
	Line  3216: I (17:02:14.191) ADC OP: read chunk 116 376
	Line  3217: I (17:02:14.216) ADC OP: read chunk 492 400
	Line  3218: I (17:02:14.241) ADC OP: read chunk 892 400
	Line  3219: I (17:02:14.266) ADC OP: read chunk 1292 308
	Line  3621: I (17:02:16.528) ADC OP: read chunk 0 92
	Line  3622: I (17:02:16.552) ADC OP: read chunk 92 388
	Line  3623: I (17:02:16.577) ADC OP: read chunk 480 400
	Line  3624: I (17:02:16.602) ADC OP: read chunk 880 400
	Line  3625: I (17:02:16.627) ADC OP: read chunk 1280 320
	Line  4027: I (17:02:18.888) ADC OP: read chunk 0 80
	Line  4028: I (17:02:18.912) ADC OP: read chunk 80 388
	Line  4029: I (17:02:18.937) ADC OP: read chunk 468 400
	Line  4030: I (17:02:18.962) ADC OP: read chunk 868 400
	Line  4031: I (17:02:18.987) ADC OP: read chunk 1268 332
	Line  4433: I (17:02:21.248) ADC OP: read chunk 0 68
	Line  4434: I (17:02:21.272) ADC OP: read chunk 68 388
	Line  4435: I (17:02:21.297) ADC OP: read chunk 456 400
	Line  4436: I (17:02:21.322) ADC OP: read chunk 856 400
	Line  4437: I (17:02:21.347) ADC OP: read chunk 1256 344
	Line  4839: I (17:02:23.618) ADC OP: read chunk 0 56
	Line  4840: I (17:02:23.642) ADC OP: read chunk 56 388
	Line  4841: I (17:02:23.667) ADC OP: read chunk 444 400
	Line  4842: I (17:02:23.692) ADC OP: read chunk 844 400
	Line  4843: I (17:02:23.717) ADC OP: read chunk 1244 356
	Line  5226: I (17:02:25.948) ADC OP: read chunk 0 44
	Line  5227: I (17:02:25.972) ADC OP: read chunk 44 388
	Line  5228: I (17:02:25.997) ADC OP: read chunk 432 400
	Line  5229: I (17:02:26.022) ADC OP: read chunk 832 400
	Line  5230: I (17:02:26.047) ADC OP: read chunk 1232 368
	Line  5604: I (17:02:28.278) ADC OP: read chunk 0 32
	Line  5605: I (17:02:28.302) ADC OP: read chunk 32 380
	Line  5606: I (17:02:28.327) ADC OP: read chunk 412 400
	Line  5607: I (17:02:28.352) ADC OP: read chunk 812 400
	Line  5608: I (17:02:28.377) ADC OP: read chunk 1212 388
	Line  6008: I (17:02:30.638) ADC OP: read chunk 0 12
	Line  6009: I (17:02:30.662) ADC OP: read chunk 12 388
	Line  6010: I (17:02:30.687) ADC OP: read chunk 400 400
	Line  6011: I (17:02:30.712) ADC OP: read chunk 800 400
	Line  6012: I (17:02:30.737) ADC OP: read chunk 1200 400
	Line  6414: I (17:02:36.942) ADC OP: read chunk 0 388
	Line  6415: I (17:02:36.967) ADC OP: read chunk 388 400
	Line  6416: I (17:02:36.992) ADC OP: read chunk 788 400
	Line  6417: I (17:02:37.017) ADC OP: read chunk 1188 400
	Line  6418: I (17:02:37.042) ADC OP: read chunk 1588 12
	Line  6803: I (17:02:39.278) ADC OP: read chunk 0 388
	Line  6804: I (17:02:39.302) ADC OP: read chunk 388 384
	Line  6805: I (17:02:39.327) ADC OP: read chunk 772 400
	Line  6806: I (17:02:39.352) ADC OP: read chunk 1172 400
	Line  6807: I (17:02:39.377) ADC OP: read chunk 1572 28
	Line  7201: I (17:02:41.618) ADC OP: read chunk 0 372
	Line  7202: I (17:02:41.642) ADC OP: read chunk 372 388
	Line  7203: I (17:02:41.667) ADC OP: read chunk 760 400
	Line  7204: I (17:02:41.692) ADC OP: read chunk 1160 400
	Line  7205: I (17:02:41.717) ADC OP: read chunk 1560 40
	Line  7597: I (17:02:43.968) ADC OP: read chunk 0 360
	Line  7598: I (17:02:43.992) ADC OP: read chunk 360 380
	Line  7599: I (17:02:44.017) ADC OP: read chunk 740 400
	Line  7600: I (17:02:44.042) ADC OP: read chunk 1140 400
	Line  7601: I (17:02:44.067) ADC OP: read chunk 1540 60
	Line  7991: I (17:02:46.318) ADC OP: read chunk 0 340
	Line  7992: I (17:02:46.342) ADC OP: read chunk 340 380
	Line  7993: I (17:02:46.367) ADC OP: read chunk 720 400
	Line  7994: I (17:02:46.392) ADC OP: read chunk 1120 400
	Line  7995: I (17:02:46.417) ADC OP: read chunk 1520 80
	Line  8386: I (17:02:48.658) ADC OP: read chunk 0 320
	Line  8391: I (17:02:48.682) ADC OP: read chunk 320 380
	Line  8392: I (17:02:48.707) ADC OP: read chunk 700 400
	Line  8393: I (17:02:48.732) ADC OP: read chunk 1100 400
	Line  8394: I (17:02:48.757) ADC OP: read chunk 1500 100
	Line  8783: I (17:02:50.988) ADC OP: read chunk 0 300
	Line  8784: I (17:02:51.012) ADC OP: read chunk 300 388
	Line  8785: I (17:02:51.037) ADC OP: read chunk 688 400
	Line  8786: I (17:02:51.062) ADC OP: read chunk 1088 400
	Line  8787: I (17:02:51.087) ADC OP: read chunk 1488 112
	Line  9188: I (17:02:53.348) ADC OP: read chunk 0 288
	Line  9189: I (17:02:53.372) ADC OP: read chunk 288 380
	Line  9190: I (17:02:53.397) ADC OP: read chunk 668 400
	Line  9191: I (17:02:53.422) ADC OP: read chunk 1068 400
	Line  9192: I (17:02:53.447) ADC OP: read chunk 1468 132
	Line  9594: I (17:02:55.708) ADC OP: read chunk 0 268
	Line  9595: I (17:02:55.732) ADC OP: read chunk 268 388
	Line  9596: I (17:02:55.757) ADC OP: read chunk 656 400
	Line  9597: I (17:02:55.782) ADC OP: read chunk 1056 400
	Line  9598: I (17:02:55.807) ADC OP: read chunk 1456 144

- issue hapens in most of the cycles. below are logs for 2 cycles showing the issue at the chunk boundary (truncated to save space)

I (17:02:02.378) ADC OP: read chunk 0 184
I (17:02:02.402) ADC OP: read chunk 184 384
I (17:02:02.427) ADC OP: read chunk 568 400
I (17:02:02.452) ADC OP: read chunk 968 400
I (17:02:02.477) ADC OP: read chunk 1368 232
I (17:02:02.478) ADC OP: 0 4 520
I (17:02:02.478) ADC OP: 1 3 24
I (17:02:02.479) ADC OP: 2 4 628
I (17:02:02.480) ADC OP: 3 3 38
I (17:02:02.481) ADC OP: 4 4 509
I (17:02:02.481) ADC OP: 5 3 27
I (17:02:02.482) ADC OP: 6 4 509
I (17:02:02.482) ADC OP: 7 3 44
I (17:02:02.483) ADC OP: 8 4 511
I (17:02:02.484) ADC OP: 9 3 25
I (17:02:02.484) ADC OP: 10 4 607
I (17:02:02.485) ADC OP: 11 3 51
I (17:02:02.486) ADC OP: 12 4 531
I (17:02:02.486) ADC OP: 13 3 29
I (17:02:02.487) ADC OP: 14 4 582
I (17:02:02.487) ADC OP: 15 3 52
I (17:02:02.488) ADC OP: 16 4 533
I (17:02:02.489) ADC OP: 17 3 33
I (17:02:02.489) ADC OP: 18 4 503
I (17:02:02.490) ADC OP: 19 3 54
I (17:02:02.491) ADC OP: 20 4 511
I (17:02:02.491) ADC OP: 21 3 40
I (17:02:02.492) ADC OP: 22 4 505
I (17:02:02.493) ADC OP: 23 3 63
I (17:02:02.493) ADC OP: 24 4 520
I (17:02:02.494) ADC OP: 25 3 48
I (17:02:02.494) ADC OP: 26 4 502
I (17:02:02.495) ADC OP: 27 3 76
I (17:02:02.496) ADC OP: 28 4 511
I (17:02:02.496) ADC OP: 29 3 60
I (17:02:02.497) ADC OP: 30 4 444
I (17:02:02.498) ADC OP: 31 3 73
I (17:02:02.498) ADC OP: 32 4 521
I (17:02:02.499) ADC OP: 33 3 57
I (17:02:02.500) ADC OP: 34 4 499
I (17:02:02.500) ADC OP: 35 3 83
I (17:02:02.501) ADC OP: 36 4 524
I (17:02:02.502) ADC OP: 37 3 60
I (17:02:02.502) ADC OP: 38 4 507
I (17:02:02.503) ADC OP: 39 3 84
I (17:02:02.503) ADC OP: 40 4 521
I (17:02:02.504) ADC OP: 41 3 61
I (17:02:02.505) ADC OP: 42 4 442
I (17:02:02.505) ADC OP: 43 3 82
I (17:02:02.506) ADC OP: 44 4 533
**_I (17:02:02.507) ADC OP: 45 3 54
I (17:02:02.507) ADC OP: 46 3 31_**
I (17:02:02.508) ADC OP: 47 4 526
I (17:02:02.509) ADC OP: 48 3 37


I (17:02:09.448) ADC OP: read chunk 0 144
I (17:02:09.472) ADC OP: read chunk 144 388
I (17:02:09.497) ADC OP: read chunk 532 400
I (17:02:09.522) ADC OP: read chunk 932 400
I (17:02:09.547) ADC OP: read chunk 1332 268
I (17:02:09.548) ADC OP: 0 4 521
I (17:02:09.549) ADC OP: 1 3 43
I (17:02:09.549) ADC OP: 2 4 507
I (17:02:09.550) ADC OP: 3 3 44
I (17:02:09.551) ADC OP: 4 4 509
I (17:02:09.551) ADC OP: 5 3 44
I (17:02:09.552) ADC OP: 6 4 504
I (17:02:09.552) ADC OP: 7 3 46
I (17:02:09.553) ADC OP: 8 4 509
I (17:02:09.554) ADC OP: 9 3 45
I (17:02:09.554) ADC OP: 10 4 524
I (17:02:09.555) ADC OP: 11 3 41
I (17:02:09.555) ADC OP: 12 4 554
I (17:02:09.556) ADC OP: 13 3 39
I (17:02:09.556) ADC OP: 14 4 559
I (17:02:09.557) ADC OP: 15 3 30
I (17:02:09.558) ADC OP: 16 4 566
I (17:02:09.558) ADC OP: 17 3 28
I (17:02:09.559) ADC OP: 18 4 534
I (17:02:09.559) ADC OP: 19 3 39
I (17:02:09.560) ADC OP: 20 4 508
I (17:02:09.561) ADC OP: 21 3 47
I (17:02:09.561) ADC OP: 22 4 569
I (17:02:09.562) ADC OP: 23 3 26
I (17:02:09.562) ADC OP: 24 4 575
I (17:02:09.563) ADC OP: 25 3 45
I (17:02:09.564) ADC OP: 26 4 556
I (17:02:09.564) ADC OP: 27 3 46
I (17:02:09.565) ADC OP: 28 4 557
I (17:02:09.566) ADC OP: 29 3 53
I (17:02:09.566) ADC OP: 30 4 550
I (17:02:09.567) ADC OP: 31 3 51
I (17:02:09.567) ADC OP: 32 4 562
I (17:02:09.568) ADC OP: 33 3 59
I (17:02:09.569) ADC OP: 34 4 555
**_I (17:02:09.569) ADC OP: 35 3 55
I (17:02:09.570) ADC OP: 36 3 34_**
I (17:02:09.571) ADC OP: 37 4 571
I (17:02:09.571) ADC OP: 38 3 22
I (17:02:09.572) ADC OP: 39 4 523
I (17:02:09.572) ADC OP: 40 3 42
I (17:02:09.573) ADC OP: 41 4 557

Here is the slightly reworked code for the read loop

   ```
...
	if(xSemaphoreTake(adc_mutex, ( TickType_t ) 100 )) // 1 sec wait
   		{
		adc_digi_start();
		while(index < NR_SAMPLES * ADC_RESULT_BYTE && ret == ESP_OK)
			{
			ret = adc_digi_read_bytes(res.result + index, b2read, &ret_num, 1000);
			if (ret == ESP_ERR_TIMEOUT || ret == ESP_ERR_INVALID_STATE)
				break;
			ESP_LOGI(TAG, "read chunk %d %d", index, ret_num);
			index += ret_num;
			b2read -= ret_num;
			}
		adc_digi_stop();
		xSemaphoreGive(adc_mutex);
		}
	else
		{
		ESP_LOGI(TAG, "cannot take adc mutex");
		esp_restart();
		}
...
    if (ret == ESP_OK)
		{
		indexd = indexc = 0;
		c_val = calloc(NR_SAMPLES, sizeof(uint16_t));
		if(!c_val)
			{
			ESP_LOGI(TAG, "not enough memory to allocate intermediate vector");
			return ESP_FAIL;
			}
		for(i = 0; i < NR_SAMPLES; i++)
			{
			ESP_LOGI(TAG, "%d %d %d", i, res.presult[i].type2.channel, res.presult[i].type2.data);
			if(res.presult[i].type2.channel == channel[0])
				c_val[indexc++] = res.presult[i].type2.data;
                      }
               }
...

```

**Comment by ves011** [comment]
Hi

Now its clear samples are lost in between chunks.
In the picture below should be a 50Hz sinus signal. between chunk 1 and 2  some samples are lost :(
These are the chunks:
```
I (50433) --------------------- ADC read loop:: 72 / 268 /268 / 0
I (50433) ADC DMA: read chunk
I (50453) --------------------- ADC read loop:: 73 / 364 /632 / 0
I (50453) ADC DMA: read chunk
I (50473) --------------------- ADC read loop:: 74 / 400 /1032 / 0
I (50473) ADC DMA: read chunk
I (50503) --------------------- ADC read loop:: 75 / 400 /1432 / 0
I (50503) ADC DMA: read chunk
I (50523) --------------------- ADC read loop:: 76 / 168 /1600 / 0
```
![Capture](https://user-images.githubusercontent.com/115264152/222228569-ddb3d33f-1b35-4776-ab3f-51b66867cae0.JPG)

**Comment by Erlkoenig90** [comment]
I am experiencing the same issue on `ESP32-S3` with ESP-IDF `v5.0.2`. Samples are missing and the samplerate is lower than expected. I think I found the cause for it:

The ADC DMA callback [restarts](https://github.com/espressif/esp-idf/blob/bfe07cbef90d50cde089f8f0695c919dd7a73d1c/components/esp_adc/adc_continuous.c#L355) the DMA transfer after each DMA run, i.e. after the DMA completed all descriptors from the linked descriptor list. If this callback is not called fast enough, the ADC will produce more samples with DMA disabled/finished before the DMA transfer is restarted, so those samples will be lost.

There is a simple solution: Turn the descriptor list into a ring/loop, by [setting the last descriptors `next` pointer](https://github.com/espressif/esp-idf/blob/bfe07cbef90d50cde089f8f0695c919dd7a73d1c/components/hal/adc_hal.c#L253) to point to the first descriptor instead to `NULL`. This will make the ADC run continuously, restarting at the beginning of the buffer after finishing.

The [call to `adc_hal_digi_start`](https://github.com/espressif/esp-idf/blob/bfe07cbef90d50cde089f8f0695c919dd7a73d1c/components/esp_adc/adc_continuous.c#L355) in the DMA callback can then simply be removed. Directly [after copying the data](https://github.com/espressif/esp-idf/blob/bfe07cbef90d50cde089f8f0695c919dd7a73d1c/components/esp_adc/adc_continuous.c#L329) from the DMA buffer to the ringbuffer, we need to make the DMA descriptor available again for DMA by setting its `owner` field to `1`. Otherwise, the DMA will trigger an error. This should probably be wrapped in its own ADC HAL function.

If the software still can't keep up with the ADC sampling, the DMA buffer will overrun, i.e. the DMA will attempt to use a descriptor with its `owner` field set to `0`. To uncover such an issue, we should catch the `DSCR_ERR` interrupt and print an error message. To do this cleanly we should add another callback to the GDMA and use it from the continuous ADC driver.

I submitted a pull request #11411 which does all of this, including new APIs for setting the `owner` field to `1` and providing a callback fro the `DSCR_ERR` interrupt. This fixes the issue for me.
